OutOfSchool/OutOfSchool.Redis/CacheService.cs

```diff
@@ -95,24 +95,71 @@ await ExecuteRedisMethod(() =>
     }
 
     public Task RemoveAsync(string key)
-        => ExecuteRedisMethod(async () => {
+        => this.ExecuteRedisMethod(() =>
+        {
             cacheLock.EnterWriteLock();
             try
             {
-                await cache.RemoveAsync(key);
+                cache.Remove(key);
             }
             finally
             {
                 cacheLock.ExitWriteLock();
             }
         });
 
+    public async Task<string> GetValueAsync(string key)
+    {
+        string returnValue = string.Empty;
+
+        await ExecuteRedisMethod(() =>
```

Comment:
I see that Enter/Exit... in many places.
Is that possible to create 2 methods to wrap that logic ?
One for read and one for write locks ?

Reply:
Is locking necessary within this service? Since `CacheService` is either a **Scoped** or **Transient** service, concurrency shouldn’t be an issue.

In my opinion, for these requirements, we could create a service that isn’t coupled with the database, acting as a temporary data storage solution.

Alternatively, we could create a **Facade** for interacting with **Redis** and a **Decorator** for caching database responses. This way, we could use the **Facade** service to store temporary data.

Reply:
Ok, we have `DraftStorageService` and `IDistributedCache` and `CacheService`. "Not that we needed all that for the trip" ([Fear and Loathing in Las Vegas](https://www.goodreads.com/quotes/59592-we-had-two-bags-of-grass-seventy-five-pellets-of-mescaline)) but obviously we could add another Facade

Reply:
> `IDistributedCache`

To implement the DraftStorageService<T> logic, I only need two methods - ReadAsync() and WriteAsync(). I could create this class as a facade over IDistributedCache. But in this case, in order for correctly reading and writing to the cache, I will need to implement the locking logic, which is already implemented in the methods of the CacheService class.
Therefore, I added the IReadWriteCacheService interface and implemented its methods in CacheService.
DraftStorageService<T> class is designed  as a facade over CacheService. This class works with entities of type T, unlike CacheService, which saves and restores information in string format.